### PR TITLE
Feature/edit logic of resending mails

### PR DESCRIPTION
**Pull request by BelehovEgor**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/662

### CODE REVIEW

File: test/MessageService.Broker.UnitTests/Helpers/EmailSenderTests.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            _mocker = new AutoMocker();
+            _sender = _mocker.CreateInstance<EmailSender>();
+
+            _mocker
+                .Setup<IOptions<SmtpCredentialsOptions>, SmtpCredentialsOptions>(x => x.Value)
+                .Returns(new SmtpCredentialsOptions
+                {
+                    Email = "lt.digitaloffice@gmail.com",
+                    Host = "smtp.gmail.com",
+                    Password = "%4fgT1_3ioR",
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply test/MessageService.Broker.UnitTests/Helpers/EmailSenderTests.cs:29]: this password we have in appsettings, how can I hide it?

File: src/MessageService/Controllers/UnsentEmailController.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        [HttpDelete("resend")]
+        public OperationResultResponse<bool> Resend(
+            [FromServices] IResendEmailCommand command,
+            [FromQuery] Guid unsentEmailId)
+        {
+            return command.Execute(unsentEmailId);
+        }
+
+        [HttpGet("getall")]
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService/Controllers/UnsentEmailController.cs:21]: find)

File: src/MessageService.Models.Dto/Models/Emails/EmailInfo.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace LT.DigitalOffice.MessageService.Models.Dto.Models.Emails
+{
+    public class EmailInfo
+    {
+        public Guid Id { get; set; }
+        public string To { get; set; }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Models.Dto/Models/Emails/EmailInfo.cs:8]: Very informative comment. Ok I will rename to "receiver"

File: src/MessageService.Data.Provider.MsSql.Ef/Migrations/20210614151600_AddUnsentEmailTable.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            migrationBuilder
+                .CreateTable(
+                    name: DbUnsentEmail.TableName,
+                    columns: table => new
+                    {
+                        Id = table.Column<Guid>(nullable: false),
+                        EmailId = table.Column<Guid>(nullable: false),
+                        CreatedAt = table.Column<DateTime>(nullable: false),
+                        LastSendAt = table.Column<DateTime>(nullable: false),
+                        TotalSendingCount = table.Column<uint>(nullable: false)
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Data.Provider.MsSql.Ef/Migrations/20210614151600_AddUnsentEmailTable.cs:23]: becouse count can't be negative

File: src/MessageService.Business/UnsentEmailCommands/Interfaces/IResendEmailCommand.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.Kernel.Responses;
+using System;
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Business/UnsentEmailCommands/Interfaces/IResendEmailCommand.cs:3]: fixed

File: src/MessageService.Broker/Helpers/EmailResender.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+using System;
+using System.Net;
+using System.Net.Mail;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.MessageService.Broker.Helpers
+{
+    public class EmailResender
+    {
+        private IUnsentEmailRepository _unsentEmailRepository;
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Broker/Helpers/EmailResender.cs:16]: fixed

File: src/MessageService.Data/Interfaces/IUnsentEmailRepository.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+{
+    [AutoInject]
+    public interface IUnsentEmailRepository
+    {
+        IEnumerable<DbUnsentEmail> GetAll();
+        IEnumerable<DbUnsentEmail> Find(int skipCount, int takeCount, out int totalCount);
+        void Add(DbUnsentEmail email);
+        DbUnsentEmail Get(Guid id);
+        bool Remove(DbUnsentEmail email);
+        void IncrementTotalCount(DbUnsentEmail email);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Data/Interfaces/IUnsentEmailRepository.cs:16]: fixed

File: src/MessageService.Broker/Helpers/EmailSender.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                    TotalSendingCount = 1
+                });
+
+            return false;
+        }
+
+        public bool ResendEmail(Guid unsentEmailId)
+        {
+            var unsentEmail = _unsentEmailRepository.Get(unsentEmailId);
+            if(Send(unsentEmail.Email))
```
Comment (djino97) [inline src/MessageService.Broker/Helpers/EmailSender.cs:97]: add empty space

File: src/MessageService.Broker/Helpers/EmailSender.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                            "Errors while sending email to {to} with subject: {subject} and body: {body}. Email replaced to resend queue.",
+                            email.Receiver,
+                            email.Subject,
+                            email.Body);
+
+                return false;
+            }
+
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Broker/Helpers/EmailSender.cs]: Fixed

File: src/MessageService.Broker/Helpers/EmailSender.cs (hunk 8 above)
Comment (BelehovEgor) [reply src/MessageService.Broker/Helpers/EmailSender.cs]: Fixed

**Review by TIove — APPROVED** [review]

**Review by djino97 — APPROVED** [review]

File: src/MessageService/Controllers/UnsentEmailController.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        [HttpDelete("resend")]
+        public OperationResultResponse<bool> Resend(
+            [FromServices] IResendEmailCommand command,
+            [FromQuery] Guid unsentEmailId)
+        {
+            return command.Execute(unsentEmailId);
+        }
+
+        [HttpGet("find")]
+        public UnsentEmailsResponse Find(
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService/Controllers/UnsentEmailController.cs:22]: What is it?

File: src/MessageService.Business/Commands/Workspace/Interfaces/ICreateWorkspaceCommand.cs [hunk 11]
```diff
@@ -1,9 +1,9 @@
 ﻿using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.Kernel.Responses;
 using LT.DigitalOffice.MessageService.Models.Dto.Requests.Workspace;
-using LT.DigitalOffice.MessageService.Models.Dto.Responses;
 using System;
 
-namespace LT.DigitalOffice.MessageService.Business.WorkspaceCommands.Interfaces
+namespace LT.DigitalOffice.MessageService.Business.Commands.Workspace.Interfaces
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Business/Commands/Workspace/Interfaces/ICreateWorkspaceCommand.cs:6]: the namespace was created based on the folder structure

File: src/MessageService.Broker/Helpers/EmailSender.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            return false;
+        }
+
+        public bool ResendEmail(Guid unsentEmailId)
+        {
+            var unsentEmail = _unsentEmailRepository.Get(unsentEmailId);
+            if (Send(unsentEmail.Email))
+            {
+                _unsentEmailRepository.Remove(unsentEmail);
+                return true;
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Broker/Helpers/EmailSender.cs:61]: added

File: src/MessageService.Broker/Helpers/EmailSender.cs [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            }
+
+            _unsentEmailRepository.IncrementTotalCount(unsentEmail);
+            return false;
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/MessageService.Broker/Helpers/EmailSender.cs:66]: added

**Review by MariyaSin — APPROVED** [review]

**Pull request merged**